### ISSUE TITLE
Sorts RoundingMode and (_ BitVec 3) are incompatible

### ISSUE THREAD

**Issue by cgcgbcbc** [Z3Prover/z3#251]
In my program this exception is raised when I call z3::solver.push().
I want to know the problem so I add a debug `std::cout << solver` before that call, the result is:
```
(declare-fun lv_f_0 () (_ FloatingPoint 8 24))
(assert true)
(assert (fp.eq (fp.mul roundNearestTiesToEven
           lv_f_0
           (fp #b0 #x69 #b00000000000000000000000))
(fp #b0 #x01 #b00000000000000000000000)))
```

Input the above formula to z3 binary, no exception raised.

But if I write a simple program for reproducing the problem, **no exception raised**: ( I manually followed my program execution and tracked all the expression creation, then extract to the following code) 
```cpp
#include<z3++.h>
#include<z3.h>

#include<iostream>
#include<string>

int main() {
    z3::context ctx;
    z3::tactic t = z3::tactic(ctx, "qfnra");

    z3::solver s = t.mk_solver();

    z3::symbol symbol(ctx, Z3_mk_string_symbol(ctx, "lv_f_0"));
    z3::sort sort(ctx, Z3_mk_fpa_sort_32(ctx));
    z3::expr lv_f_0(ctx, Z3_mk_const(ctx, symbol, sort));

    std::string hex_a = "0x3E90000000000000";
    uint64_t bv_a = std::stoull(hex_a, 0, 16);
    double d_a = * ((double *) & bv_a);
    z3::sort sort32(ctx, Z3_mk_fpa_sort_32(ctx));
    z3::expr a(ctx, Z3_mk_fpa_numeral_float(ctx, d_a, sort32));

    z3::ast rne(ctx, Z3_mk_fpa_rne(ctx));

    z3::expr mul(ctx, Z3_mk_fpa_mul(ctx, rne, lv_f_0, a));

    std::string hex_b = "0x3810000000000000";
    uint64_t bv_b = std::stoull(hex_b, 0, 16);
    double d_b = * ((double *) & bv_b);
    z3::expr b(ctx, Z3_mk_fpa_numeral_float(ctx, d_b, sort32));

    z3::expr to_check(ctx, Z3_mk_fpa_eq(ctx, mul, b));

    s.add(to_check);

    s.push();

    std::cout << s << std::endl;

    std::cout << s.check() << std::endl;

    return 0;
}
```

I'm not sure if this is a z3 bug, may be my real program has some other hidden own bugs, I tried to debug , but met some trouble: even I compile z3 with debug mode, I still couldn't step into api calls..

**Comment by cgcgbcbc** [comment]
[SO link](http://stackoverflow.com/questions/33217215/sorts-roundingmode-and-bitvec-3-are-incompatible)

**Comment by wintersteiger** [comment]
This is a bug somewhere in the floating point theory, but this test case doesn't trigger it. 

Note that you're trying to use the `qfnra` tactic, which does not support flloating point numbers, it only supports reals. Therefore, the real bug here is that Z3 doesn't throw an error much earlier.

**Comment by wintersteiger** [comment]
> even I compile z3 with debug mode, I still couldn't step into api calls..

This depends on your system. If you're using a .NET language then you need to "Enable native code debugging" in your project properties. If you're using C++ from Visual Studio, you have to make sure that it can find the sources of Z3; it should be fine though if you generate the project file in debug mode (add --vs to python scripts\mk_make.py) and then add that to your solution.

**Comment by cgcgbcbc** [comment]
@wintersteiger  finally I can construct a witness for the bug:
```
#include <iostream>
#include <string>

#include <z3++.h>
#include <z3.h>

void error_handler(Z3_context ctx, Z3_error_code code) {
    if (code != Z3_OK) {
    }
}

int main() {
    z3::config cfg;
    cfg.set("MODEL", true);
    cfg.set("PROOF", true);
    z3::context c(cfg);
    z3::solver s(c);
    Z3_set_error_handler(c, &error_handler);

    auto v0 = c.bool_val(true);
    s.push();
    s.add(v0);
    z3::ast round_mode(c, Z3_mk_fpa_rne(c));
    z3::symbol symbol(c, Z3_mk_string_symbol(c, "lv_f_0"));
    Z3_sort sort = Z3_mk_fpa_sort_32(c);
    z3::expr lv_f_0(c, Z3_mk_const(c, symbol, z3::sort(c, sort)));
    
    std::string hex = "0x3E90000000000000";
    uint64_t bv_value = std::stoull(hex, 0, 16);
    double d_value = * ((double *) & bv_value);
    z3::sort to_sort(c, Z3_mk_fpa_sort_32(c));
    z3::expr a(c, Z3_mk_fpa_numeral_float(c, d_value, to_sort));

    z3::expr mul(c, Z3_mk_fpa_mul(c, round_mode, lv_f_0, a));

    hex = "0x3810000000000000";
    bv_value = std::stoull(hex, 0, 16);

    z3::sort to_sort2(c, Z3_mk_fpa_sort_32(c));
    z3::expr b(c, Z3_mk_fpa_numeral_float(c, d_value, to_sort2));

    z3::expr to_check(c, Z3_mk_fpa_eq(c, mul, b));

    s.add(to_check);

    z3::expr result = c.bool_val(false);

    s.push();
}
```
The last push will trigger the bug.

**Comment by cgcgbcbc** [comment]
If I comment the following lines, everything is ok.
```
cfg.set("MODEL", true);
cfg.set("PROOF", true);
```

**Comment by wintersteiger** [comment]
Fantastic, thanks for the test case, this should make it a lot easier to track down!

**Comment by wintersteiger** [comment]
This problem was due to some parts of the tactics used for this problem not supporting proof generation. I've fixed those problems (as of cbf8bd8de12c593818d4dcef27cf55865aa6188f) and now it doesn't throw any exceptions and solves the problem correctly, both via the SMT2 frontend or the API.
